### PR TITLE
vkdf/buffer: map/unmap using sized aligned to memory requirements ali…

### PR DESCRIPTION
**Pull request by infapi00**
…gment

To prevent Validation Layer error VUID-VkMappedMemoryRange-size-01390.

Right now vkdf-create-buffer is allocating what the buffer really
requires. So for example, if you are trying to allocate 96 bytes, but
the specific driver allocates based on an aligment of 256, the
allocate size will be 256.

So then, if you try to fill up the whole buffer using an array of 96,
if you map/unmap only the size, you would get the validation error
below. This commit maps/unmaps using an aligned size. The copied
memory still uses the original size.

Full text of the validation error:

VKDF ERROR: validation layer: [ VUID-VkMappedMemoryRange-size-01390 ]
vkFlushMappedMemoryRanges: Size in pMemRanges[0] is 0x60, which is not
a multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize (0x100). The
Vulkan spec states: If size is not equal to VK_WHOLE_SIZE, size must
either be a multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize,
or offset plus size must equal the size of
memory. (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkMappedMemoryRange-size-01390)

### CODE REVIEW

File: framework/vkdf-buffer.cpp [hunk 1]
```diff
@@ -54,12 +60,14 @@ vkdf_buffer_map_and_fill(VkdfContext *ctx,
    assert(buf.mem_props & VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT);
 
    void *mapped_memory;
-   vkdf_memory_map(ctx, buf.mem, offset, size, &mapped_memory);
+   VkDeviceSize mapping_size = align64(size, buf.mem_reqs.alignment);
```
Comment (itoral) [inline framework/vkdf-buffer.cpp]: let's call this 'aligned_size'

File: framework/vkdf-buffer.cpp [hunk 2]
```diff
@@ -54,12 +60,14 @@ vkdf_buffer_map_and_fill(VkdfContext *ctx,
...
+
```
Comment (itoral) [inline framework/vkdf-buffer.cpp]: remove the blank line here

File: framework/vkdf-buffer.cpp [hunk 3]
```diff
@@ -79,7 +87,9 @@ vkdf_buffer_map_and_fill_elements(VkdfContext *ctx,
    VkDeviceSize size = num_elements * dst_stride;
    assert(buf.mem_reqs.size >= size);
 
-   vkdf_memory_map(ctx, buf.mem, offset, size, &mapped_memory);
+   VkDeviceSize mapping_size = align64(size, buf.mem_reqs.alignment);
+
```
Comment (itoral) [inline framework/vkdf-buffer.cpp]: Same comments as above

File: framework/vkdf-buffer.cpp [hunk 4]
```diff
@@ -44,6 +44,12 @@ vkdf_create_buffer(VkdfContext *ctx,
    return buffer;
 }
 
+static inline uint64_t
+align64(uint64_t value, unsigned alignment)
+{
+   return (value + alignment - 1) & ~((uint64_t)alignment - 1);
```
Comment (itoral) [inline framework/vkdf-buffer.cpp:57]: This only works for values of alignment that are powers of 2, so an assert would be useful (the assert could check that only one bit in alignment is set)

File: framework/vkdf-buffer.cpp [hunk 5]
```diff
@@ -90,7 +100,7 @@ vkdf_buffer_map_and_fill_elements(VkdfContext *ctx,
       src_offset += src_stride;
    }
 
-   vkdf_memory_unmap(ctx, buf.mem, buf.mem_props, offset, size);
+   vkdf_memory_unmap(ctx, buf.mem, buf.mem_props, offset, mapping_size);
 }
 
 void
```
Comment (itoral) [inline framework/vkdf-buffer.cpp:111]: One other comment: change the shortlog to this:

"framework/buffer: align buffer size for map/unmap to honor memory requirements"

**Comment by infapi00** [comment]
All changes suggested done. Just pushed again.

**Pull request merged**